agent/jvm/src/main/java/org/jolokia/jvmagent/JvmAgent.java

```diff
@@ -73,29 +78,60 @@ public static void premain(String agentArgs) {
      *
      * @param agentArgs arguments as given on the command line
      */
-    public static void agentmain(String agentArgs) {
+    public static void agentmain(String agentArgs, Instrumentation instrumentation) {
         JvmAgentConfig config = new JvmAgentConfig(agentArgs);
         if (!config.isModeStop()) {
-            startAgent(config,false);
+            startAgent(config,false, instrumentation);
         } else {
             stopAgent();
         }
     }
 
-    private static void startAgent(JvmAgentConfig pConfig,boolean pLazy)  {
-        try {
-            server = new JolokiaServer(pConfig,pLazy);
+    private static void startAgent(final JvmAgentConfig pConfig, final boolean pLazy, final Instrumentation instrumentation)  {
+        // start the JolokiaServer in a new daemon thread 
+        Thread jolokiaStartThread = new Thread("JolokiaStart") {
+            public void run() {
+                try {
+                    // block until the server supporting early detection is initialized
+                    awaitServerInitialization(pConfig, instrumentation);
+                    
+                    server = new JolokiaServer(pConfig,pLazy);
 
-            server.start();
-            setStateMarker();
+                    server.start();
+                    setStateMarker();
 
-            System.out.println("Jolokia: Agent started with URL " + server.getUrl());
-        } catch (RuntimeException exp) {
-            System.err.println("Could not start Jolokia agent: " + exp);
-        } catch (IOException exp) {
-            System.err.println("Could not start Jolokia agent: " + exp);
-        }
+                    System.out.println("Jolokia: Agent started with URL " + server.getUrl());
+                } catch (RuntimeException exp) {
+                    System.err.println("Could not start Jolokia agent: " + exp);
+                } catch (IOException exp) {
+                    System.err.println("Could not start Jolokia agent: " + exp);
+                }  
+            }
+        };
+        jolokiaStartThread.setDaemon(true);
+        jolokiaStartThread.start();
     }
+    
+    /**
+     * Lookup the server detectors and await the server initialization.
+     *    
+     * @param instrumentation 
+     * @see ServerDetector#earlyDetect(Instrumentation)
+     * @see ServerDetector#awaitServerInitialization(Instrumentation)
+     */
+    private static void awaitServerInitialization(JvmAgentConfig pConfig, final Instrumentation instrumentation) {
+        List<ServerDetector> detectors = MBeanServerHandler.lookupDetectors();
+        for (ServerDetector detector:detectors) {
+            // TODO add feature to JvmAgentConfig to be able to skip earlyDetect & awaitServerInitialization 
+            // for a detector. 
+            if (detector.earlyDetect(instrumentation)) {
```

Comment:
I wonder whether we could combine `earlyDetect` and `awaitServerInitialization` into a single method e.g. `serverPreInit`. The JBossServer detector could do the detection and waiting within this method, so we could remove one method from `ServerDetector`.

On the other end, we could use `earlyDetect()` to start only a thread if a detector returns true. I'm not sure whether there could be side effects when the Jolokia JVM HttpServer and the normal initialization continue in parallel (or when an external requests comes in early). But I guess that should be no problem, so I tend to reduce this to a single method.